src/IO/S3/tests/TestPocoHTTPServer.h

```diff
@@ -0,0 +1,68 @@
+#pragma once
+
+#include <cstddef>
+#include <memory>
+#include <random>
+#include <string>
+
+#include <Poco/Net/HTTPRequestHandler.h>
+#include <Poco/Net/HTTPRequestHandlerFactory.h>
+#include <Poco/Net/HTTPServer.h>
+#include <Poco/Net/HTTPServerParams.h>
+#include <Poco/Net/NetException.h>
+#include <Poco/Net/ServerSocket.h>
+#include <Poco/AutoPtr.h>
+#include <Poco/SharedPtr.h>
+
+
+template <typename RequestHandler>
+class HTTPRequestHandlerFactory : public Poco::Net::HTTPRequestHandlerFactory
+{
+    virtual Poco::Net::HTTPRequestHandler * createRequestHandler(const Poco::Net::HTTPServerRequest &) override
+    {
+        return new RequestHandler();
+    }
+
+public:
+    virtual ~HTTPRequestHandlerFactory() override
+    {
+    }
+};
+
+template <typename RequestHandler>
+class TestPocoHTTPServer
+{
+    std::unique_ptr<Poco::Net::ServerSocket> server_socket;
+    std::unique_ptr<Poco::Net::HTTPServer> server;
+    Poco::SharedPtr<HTTPRequestHandlerFactory<RequestHandler>> handler_factory;
+    Poco::AutoPtr<Poco::Net::HTTPServerParams> server_params;
+
+public:
+    TestPocoHTTPServer():
+        handler_factory(new HTTPRequestHandlerFactory<RequestHandler>()),
+        server_params(new Poco::Net::HTTPServerParams())
+    {
+        std::random_device device;
+        std::mt19937 generator(device());
+        std::uniform_int_distribution<> distribution(10000, 60000);
+        for (int attempt = 100; attempt >= 0; --attempt)
+        {
+            try
+            {
+                server_socket = std::make_unique<Poco::Net::ServerSocket>(distribution(generator));
```

Comment:
If you pass 0 as port, the port will be determined automatically

Reply:
Yeah, it works.